### ISSUE TITLE
Abstract Base Class and Inheritance

### ISSUE THREAD

**Issue by falahati** [RSuter/NSwag#1652]
Hey,

I have a special situation in which one of the methods of my controller returns an array of an abstract base class which contains children of that class. However, the code generated with the NSwag Studio is unusable and also there is no way to access properties exclusive to the child classes even if the code generated by the NSwag is fixed manually. Let's take a look at the following examples:

**What Happens Now:**

Api Server:

```C#
public abstract class BaseSettings {
    public string SharedProperty { get; set; } = "Shared";
}
public class BarSettings : BaseSettings {
    public string BarProperty { get; set; } = "Bar";
}
public class FooSettings : BaseSettings {
    public string FooProperty { get; set; } = "Foo";
}

public class SampleController : ApiController {
    [SwaggerResponse(HttpStatusCode.OK, typeof(BaseSettings[]))]
    public IEnumerable<BaseSettings> GetSettings() {
        return new BaseSettings[] { new BarSettings(), new FooSettings()  };
    }
}
```

Generated Definition:

```JSON
.......
"BaseSettings": {
      "type": "object",
      "x-abstract": true,
      "additionalProperties": false,
      "properties": {
        "SharedProperty": {
          "type": "string"
         }
        }
      },
.......
    "/api/Settings/GetSettings": {
      "get": {
        "tags": [
          "Settings"
        ],
        "operationId": "Settings_GetSettings",
        "responses": {
          "200": {
            "x-nullable": true,
            "description": "",
            "schema": {
              "type": "array",
              "items": {
                "$ref": "#/definitions/BaseSettings"
              }
            }
          }
        },
.......
```

Actual Response;

```JSON
[
   {
      "SharedProperty": "Shared",
      "BarProperty": "Bar"
    },
    {
      "SharedProperty": "Shared",
      "FooProperty": "Foo"
    }
]
```

Generated client code:

```C#
    [System.CodeDom.Compiler.GeneratedCode("NJsonSchema", "9.10.67.0 (Newtonsoft.Json v9.0.0.0)")]
    public partial abstract class BaseSettings : System.ComponentModel.INotifyPropertyChanged
    {
        private string _sharedProperty;
        public event System.ComponentModel.PropertyChangedEventHandler PropertyChanged;
    
        [Newtonsoft.Json.JsonProperty("SharedProperty", Required = Newtonsoft.Json.Required.Default, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
        public string SharedProperty
        {
            get { return _sharedProperty; }
            set 
            {
                if (_sharedProperty != value)
                {
                    _sharedProperty = value; 
                    RaisePropertyChanged();
                }
            }
        }
    
        public string ToJson() 
        {
            return Newtonsoft.Json.JsonConvert.SerializeObject(this);
        }
        
        public static BaseSettings FromJson(string data)
        {
            return Newtonsoft.Json.JsonConvert.DeserializeObject<BaseSettings>(data);
        }

        protected virtual void RaisePropertyChanged([System.Runtime.CompilerServices.CallerMemberName] string propertyName = null)
        {
            var handler = PropertyChanged;
            if (handler != null) 
                handler(this, new System.ComponentModel.PropertyChangedEventArgs(propertyName));
        }    
    }
```

Now as you can clearly see from the above example,
1. This method by itself is not executable as the generated `BaseSettings` class is, in fact, an `abstract` class and can't instantiate from. This can be solved by manually removing `abstract` modifier from the generated C# client code tho.

2. However, the problem with not being able to access child properties remains. This is especially unfortunate because the child class's property values are in fact available in the response generated by WebAPI, but we can't decode it right.

Is there any way to solve this or at least a way to work around this problem?

**Proposed solution:**

* A new property for the `SwaggerResponseAttribute` class allowing us to provide a list of additional types for the WebApi Server:

```C#
public class SampleController : ApiController {
    [SwaggerResponse(HttpStatusCode.OK, typeof(BaseSettings[]), AdditionalTypes = new [] {typeof(BarSettings[]), typeof(FooSettings[])})]
    public IEnumerable<BaseSettings> GetSettings() {
        return new BaseSettings[] { new BarSettings(), new FooSettings()  };
    }
}
```

* Generating type definition for the child classes based on this property; also containing an `x-inherits` property (`x-implements` for interfaces?):

```JSON
.......
"BaseSettings": {
      "type": "object",
      "x-abstract": true,
      "additionalProperties": false,
      "properties": {
        "SharedProperty": {
          "type": "string"
         }
        }
      },
"BarSettings": {
      "type": "object",
      "x-abstract": false,
      "x-inherits": "#/definitions/BaseSettings",
      "additionalProperties": false,
      "properties": {
        "SharedProperty": {
          "type": "string"
         },
        "BarProperty": {
          "type": "string"
         }
        }
      },
"FooSettings": {
      "type": "object",
      "x-abstract": false,
      "x-inherits": "#/definitions/BaseSettings",
      "additionalProperties": false,
      "properties": {
        "SharedProperty": {
          "type": "string"
         },
        "FooProperty": {
          "type": "string"
         }
        }
      },
.......
```

* Specifying the target type in the method response:

```JSON
[
   {
      "$type": "#/definitions/BarSettings",
      "SharedProperty": "Shared",
      "BarProperty": "Bar"
    },
    {
      "$type": "#/definitions/FooSettings",
      "SharedProperty": "Shared",
      "FooProperty": "Foo"
    }
]
```

I suspect this problem to arise also when an interface is used instead of an abstract class.

**Comment by RSuter** [comment]
Did you check: https://github.com/RSuter/NJsonSchema/wiki/Inheritance

**Comment by RSuter** [comment]
Can we close? Or didnt this answer your questions?

**Comment by ranouf** [comment]
Hi,

About this page: https://github.com/RSuter/NJsonSchema/wiki/Inheritance

At the end there is:
`[JsonInheritance("Dog", typeof(Dog))]`

And at the beginning there is:
[KnownType(typeof(Dog))]

I try to find in your repository JsonInheritance implementation but I didnt, I tried to find on google, nothing appears. 

KnownType worked but the discriminator is not what I expected, I m supposed to something like MyItemDto, but I only see Interface`1.

I m not sure what to do. Could you help me?

**Comment by falahati** [comment]
Sorry for the late response,

I managed to mark my generated swagger file using the `KnowType`, `JsonInheritanceConverter` and piece of code I found here at #1581 which solves the second problem I mentioned here, kinda.

However the first problem still stands, the one that @ranouf  specifically asked for a solution at #1766 .

The solution, however, is obvious, considering all `abstract` definitions as an `interface` instead of a `class`. This is more flexible as it also allows for multi inherited (or multi implementation) situations. In my project, for example, some objects implement multiple interfaces, therefore linear inheritance with `abstract class` is not an option for the generated code. I suppose this should also help with `anyOf`, `allOf` and `oneOf` rules.

That being said, the bigger problem, for now, is not even creating `interface`s instead of `abstract class`es; it is the fact **that the generator won't define inheritance** at all for child classes.

In fact, I am going to take a look at the library to see if I can think of something, last night it took me one complete hour to find out that the code for generating schema files are not in fact in this library but in the NJSchema. GitHub is not very kind with big projects, should have used a proper IDE to check the code.
